components/ws-manager/pkg/manager/manager.go

```diff
@@ -629,6 +657,57 @@ func podRunning(clientset client.Client, podName, namespace string) wait.Conditi
 	}
 }
 
+func isPodUnschedulable(clientset client.Client, podName, namespace string) bool {
+	var pod corev1.Pod
+	err := clientset.Get(context.Background(), types.NamespacedName{Namespace: namespace, Name: podName}, &pod)
+	if err != nil {
+		return false
+	}
+
+	if pod.Status.Phase != corev1.PodPending {
+		return false
+	}
+
+	for _, c := range pod.Status.Conditions {
+		if c.Type != corev1.PodScheduled {
+			continue
+		}
+
+		if c.Status == corev1.ConditionFalse && c.Reason == corev1.PodReasonUnschedulable {
+			return true
+		}
+	}
+
+	return false
+}
+
+func deleteWorkspacePod(clientset client.Client, name, namespace string) error {
```

Comment:
Since it's brutal to remove the workspace pod by removing the finalizer, consider renaming it. For example, `deleteWorkspacePodForce`.